modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -1082,44 +1105,93 @@
           [handle-filter
            handle-breakout+aggregation
            handle-order-by
            handle-fields
            handle-limit
            handle-page]))
 
+(s/defn ^:private generate-aggregation-pipeline :- {:projections Projections, :query Pipeline}
+  "Generate the aggregation pipeline. Returns a sequence of maps representing each stage."
+  [inner-query :- mbql.s/MBQLQuery]
+  (add-aggregation-pipeline inner-query {:projections [], :query []}))
+
 (defn- query->collection-name
   "Return `:collection` from a source query, if it exists."
   [query]
   (mbql.u/match-one query
     (_ :guard (every-pred map? :collection))
     ;; ignore source queries inside `:joins` or `:collection` outside of a `:source-query`
-    (when (and (= (last &parents) :source-query)
-               (not (contains? (set &parents) :joins)))
+    (when (let [parents (set &parents)]
+            (and (contains? parents :source-query)
+                 (not (contains? parents :joins))))
       (:collection &match))))
 
 (defn- log-aggregation-pipeline [form]
   (when-not qp.i/*disable-qp-logging*
     (log/tracef "\nMongo aggregation pipeline:\n%s\n"
                 (u/pprint-to-str 'green (walk/postwalk #(if (symbol? %) (symbol (name %)) %) form)))))
 
+(defn simple-mbql->native
+  "Compile a simple (non-nested) MBQL query."
+  [query]
+  (let [{proj :projections, generated-pipeline :query}
+        (generate-aggregation-pipeline (or (:query query) query))]
+    {:projections proj
+     :query       generated-pipeline}))
+
+(defn parse-query-string
+  "Parse a serialized native query. Like a normal JSON parse, but handles BSON/MongoDB extended JSON forms."
+  [^String s]
+  (try
+    (mapv (fn [^org.bson.BsonValue v] (-> v .asDocument com.mongodb.BasicDBObject.))
+          (org.bson.BsonArray/parse s))
+    (catch Throwable e
+      (throw (ex-info (tru "Unable to parse query: {0}" (.getMessage e))
+               {:type  qp.error-type/invalid-query
+                :query s}
+               e)))))
+
+(defn- mbql->native-rec
+  "Compile a potentially nested MBQL query."
+  [inner-query]
+  (if-let [source-query (-> inner-query :source-query)]
+    (let [compiled (or (when-let [nq (:native source-query)]
+                         (cond
+                           (string? nq)
+                           (-> source-query
+                               (dissoc :native)
+                               (assoc :query (parse-query-string nq)))
+
+                           :else
+                           nq))
+                       (mbql->native-rec source-query))
+          field-mappings (zipmap (mapcat #(% source-query) [:fields :breakout :aggregation])
+                                 (:projections compiled))]
+      (binding [*field-mappings* field-mappings]
+        (merge compiled (add-aggregation-pipeline inner-query compiled))))
+    (simple-mbql->native inner-query)))
+
+(defn- preprocess
+  [inner-query]
+  (add/add-alias-info inner-query))
```

Comment:
This should probably be a middleware.

Reply:
Only if it's duplicated somewhere else already.